src/ImageSharp/Processing/Normalization/HistogramEqualizationProcessor.cs

```diff
@@ -0,0 +1,120 @@
+
+using System;
+using SixLabors.ImageSharp.Advanced;
+using SixLabors.ImageSharp.PixelFormats;
+using SixLabors.ImageSharp.Processing.Processors;
+using SixLabors.Memory;
+using SixLabors.Primitives;
+
+namespace SixLabors.ImageSharp.Processing.Normalization
+{
+    /// <summary>
+    /// Applies a global histogram equalization to the image.
+    /// </summary>
+    /// <typeparam name="TPixel">The pixel format.</typeparam>
+    internal class HistogramEqualizationProcessor<TPixel> : ImageProcessor<TPixel>
+        where TPixel : struct, IPixel<TPixel>
+    {
+        /// <inheritdoc/>
+        protected override void OnFrameApply(ImageFrame<TPixel> source, Rectangle sourceRectangle, Configuration configuration)
+        {
+            var rgb48 = default(Rgb48);
+            var rgb24 = default(Rgb24);
+            MemoryAllocator memoryAllocator = configuration.MemoryAllocator;
+            int numberOfPixels = source.Width * source.Height;
+            bool is16bitPerChannel = typeof(TPixel) == typeof(Rgb48) || typeof(TPixel) == typeof(Rgba64);
+
+            // build the histogram of the grayscale levels
+            int luminanceLevels = is16bitPerChannel ? 65536 : 256;
+            Span<int> histogram = memoryAllocator.Allocate<int>(luminanceLevels, clear: true).GetSpan();
+            for (int y = 0; y < source.Height; y++)
+            {
+                Span<TPixel> row = source.GetPixelRowSpan(y);
+                for (int x = 0; x < source.Width; x++)
+                {
+                    TPixel sourcePixel = row[x];
+                    int luminance = this.GetLuminance(sourcePixel, is16bitPerChannel, ref rgb24, ref rgb48);
+                    histogram[luminance]++;
+                }
+            }
+
+            // calculate the cumulative distribution function (which will be the cumulative histogram)
+            Span<int> cdf = memoryAllocator.Allocate<int>(luminanceLevels, clear: true).GetSpan();
+            int histSum = 0;
+            for (int i = 0; i < histogram.Length; i++)
+            {
+                histSum += histogram[i];
+                cdf[i] = histSum;
+            }
+
+            // get the first none zero value of the cumulative histogram
+            int cdfMin = 0;
+            for (int i = 0; i < histogram.Length; i++)
+            {
+                if (histogram[i] != 0)
+                {
+                    cdfMin = cdf[i];
+                    break;
+                }
+            }
+
+            // creating the lookup table: subtracting cdf min, so we do not need to do that inside the for loop
+            for (int i = 0; i < histogram.Length; i++)
+            {
+                cdf[i] = cdf[i] - cdfMin;
+            }
+
+            // apply the cdf to each pixel of the image
+            double numberOfPixelsMinusCdfMin = (double)(numberOfPixels - cdfMin);
+            int luminanceLevelsMinusOne = luminanceLevels - 1;
+            for (int y = 0; y < source.Height; y++)
```

Comment:
as above, as this doesn't actually depend on the relative position of the pixel then this can be done is a single loop over all the pixels.